src/factories/guides.ts

```diff
@@ -0,0 +1,160 @@
+import { Container } from 'pixi.js'
+import { DEFAULT_GUIDES_COUNT, DEFAULT_GUIDES_MIN_GAP } from '@/consts'
+import { GuideFactory, guideFactory } from '@/factories/guide'
+import { FormatDate } from '@/models/guides'
+import { LayoutSettings } from '@/models/layout'
+import { waitForViewport } from '@/objects'
+import { emitter } from '@/objects/events'
+import { waitForScale } from '@/objects/scale'
+import { repeat } from '@/utilities/repeat'
+import { formatDateFns, labelFormats, timeIncrements } from '@/utilities/timeIncrements'
+
+const visibleGuideBoundsMargin = 300
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export async function guidesFactory() {
+  const viewport = await waitForViewport()
+  const scale = await waitForScale()
+
+  const element = new Container()
+  const guides = new Map<number, GuideFactory>()
+
+  let paused = false
+
+  let currentIncrement = 0
+  let currentAnchor = 0
+  let labelFormatter: FormatDate = (date) => date.toLocaleTimeString()
+
+  emitter.on('viewportDateRangeUpdated', () => {
+    update()
+  })
+  emitter.on('layoutCreated', (layout) => onLayoutUpdate(layout))
+  emitter.on('layoutUpdated', (layout) => onLayoutUpdate(layout))
+
+  function render(): void {
+    createGuides()
+  }
+
+  async function createGuides(): Promise<void> {
+    const guideIndexes = Array.from({ length: DEFAULT_GUIDES_COUNT }, (val, i) => i)
+
+    for await (const guideIndex of guideIndexes) {
+      await createGuide(guideIndex)
+    }
+  }
+
+  async function createGuide(index: number): Promise<void> {
+    if (guides.has(index)) {
+      return
+    }
+
+    const response = await guideFactory()
+
+    element.addChild(response.element)
+
+    guides.set(index, response)
+  }
+
+  function update(): void {
+    if (paused) {
+      return
+    }
+
+    const left = scale.invert(viewport.left - visibleGuideBoundsMargin)
+    const gapDate = scale.invert(viewport.left - visibleGuideBoundsMargin + DEFAULT_GUIDES_MIN_GAP / viewport.scale.x) as Date
```

Comment:
nit: You're already type guarding `left` might as well type guard `gapDate` as well rather than using an `as`